File: loopx/control_plane/quota/refresh_external_delivery.ts
/** Per-settlement pause acknowledgement; never a provider permission grant. */
import { createHash } from "node:crypto";
import type { JsonObject, SettlementIdentity } from "../effect_program.ts";
import { EffectRuntimeRequestError } from "../effect_runtime_errors.ts";
import { jsonObject, requireJsonObject } from "../runtime_decode.ts";

export const EXTERNAL_DELIVERY_SCHEMA = "refresh_external_delivery_v0";
export const EXTERNAL_DELIVERY_EVENT = "refresh_external_delivery";
export interface ExternalDeliveryRequest {
  suppress: boolean;
  resume_key: string | null;
}

export function decodeExternalDelivery(value: unknown): ExternalDeliveryRequest | null {
  if (value == null) return null; // Existing non-delivery callers remain local.
  const input = requireJsonObject(value, "external_delivery");
  if (typeof input.suppress !== "boolean" ||
      (input.resume_key !== null &&
       (typeof input.resume_key !== "string" || !/^[a-f0-9]{64}$/.test(input.resume_key))) ||
      (input.suppress && input.resume_key !== null)) {
    throw new EffectRuntimeRequestError("external_delivery requires suppress and a mutually exclusive resume key");
  }
  return { suppress: input.suppress, resume_key: input.resume_key };
}

function pauseKey(identity: SettlementIdentity, precedingEvent: JsonObject | null): string {
  return createHash("sha256").update(JSON.stringify([
    EXTERNAL_DELIVERY_SCHEMA, identity.effect_id, precedingEvent?.event_id ?? null,
  ])).digest("hex");
}

export function refreshExternalDelivery(
  request: ExternalDeliveryRequest | null,
  identity: SettlementIdentity,
  events: readonly JsonObject[],
  deliveryStage: boolean,
): JsonObject {
  let previous: JsonObject | null = null;
  let state: "paused" | "ready" | null = null;
  let key: string | null = null;
  let invalid = false;
  for (const event of events) {
    if (event.event_kind !== EXTERNAL_DELIVERY_EVENT || event.goal_id !== identity.goal_id ||
        event.agent_id !== identity.agent_id || event.run_id !== identity.turn_instance_id ||
        (event.todo_id ?? null) !== identity.todo_id) continue;
    const details = jsonObject(event.details);
    if ((details?.replan_obligation_id ?? null) !== identity.replan_obligation_id) continue;
    if (!details || details.schema_version !== EXTERNAL_DELIVERY_SCHEMA ||
        details.settlement_effect_id !== identity.effect_id || typeof event.event_id !== "string" ||
        (details.state !== "paused" && details.state !== "ready") ||
        typeof details.resume_key !== "string" || !/^[a-f0-9]{64}$/.test(details.resume_key)) {
      invalid = true;
      break;
    }
    if (event.event_id === previous?.event_id) continue;
    if (details.state === "paused"
      ? state === "paused" || details.resume_key !== pauseKey(identity, previous)
      : state !== "paused" || details.resume_key !== key) {
      invalid = true;
      break;
    }
    state = details.state;
    key = details.resume_key;
    previous = event;
  }
  const result = (authorized: boolean, reason: string, error: string | null = null,
    transition: JsonObject | null = null): JsonObject => ({
    schema_version: EXTERNAL_DELIVERY_SCHEMA, authorized, reason,
    error_code: error, resume_key: key, transition,
  });
  // No send phase: do not reject a local closeout or consume a resume acknowledgement.
  if (!request || (!deliveryStage && !request.suppress)) return result(false, "local_recovery");
  if (invalid) return request.suppress ? result(false, "suppressed")
    : result(false, "invalid_pause_history", "external_delivery_history_invalid");
  const transition = (next: "paused" | "ready", resumeKey: string): JsonObject => ({
    schema_version: EXTERNAL_DELIVERY_SCHEMA, settlement_effect_id: identity.effect_id,
    replan_obligation_id: identity.replan_obligation_id, state: next, resume_key: resumeKey,
  });
  if (request.suppress) {
    if (state === "paused") return result(false, "suppressed");
    key = pauseKey(identity, previous);
    return result(false, "suppressed", null, transition("paused", key));
  }
  if (request.resume_key !== null) {
    if (key === null || request.resume_key !== key) {
      return result(false, "resume_key_mismatch", "external_delivery_resume_mismatch");
    }
    return result(true, "resume_acknowledged", null,
      state === "paused" ? transition("ready", key) : null);
  }
  if (state === "paused") {
    return result(false, "resume_confirmation_required", "external_delivery_resume_required");
  }
  return result(true, state === "ready" ? "resumed" : "legacy_per_call");
}

File: loopx/control_plane/quota/refresh_external_delivery.py
"""Persist the typed resume decision inside the existing refresh serialization lock."""
from pathlib import Path
from typing import Any

from ...rollout_event_log import append_rollout_event, build_rollout_event, rollout_event_log_path
from .settlement import QuotaSettlementReadback, settlement_result_payload


def finish_external_delivery_refresh(
    payload: dict[str, Any], readback: QuotaSettlementReadback | None,
    runtime_root: Path, *, dry_run: bool,
) -> dict[str, Any]:
    if readback is None:
        return payload
    plan = readback.external_delivery
    if not isinstance(plan, dict) or plan.get("schema_version") != "refresh_external_delivery_v0":
        raise RuntimeError("TypeScript refresh external delivery result missing or invalid")
    payload["external_delivery"] = {k: v for k, v in plan.items() if k != "transition"}
    payload["external_sink_delivery_authorized"] = plan["authorized"] is True
    transition = plan.get("transition")
    if payload.get("ok") and transition and not dry_run:
        identity = readback.identity.value
        if identity is None:
            raise RuntimeError("external delivery transition has no settlement identity")
        append_rollout_event(
            rollout_event_log_path(runtime_root, identity.goal_id),
            build_rollout_event(
                goal_id=identity.goal_id, agent_id=identity.agent_id,
                todo_id=identity.todo_id, run_id=identity.turn_instance_id,
                event_kind="refresh_external_delivery", status=transition["state"],
                summary="Refresh external delivery preference recorded.", details=transition,
            ),
        )
        plan["transition"] = None  # The planned journal effect was committed once.
    return payload


def refresh_recovery_payload(
    readback: QuotaSettlementReadback, *, registry_path: Path,
    runtime_root: Path, goal_id: str, dry_run: bool,
) -> dict[str, Any] | None:
    recovery = readback.refresh_recovery
    identity = readback.identity.value
    plan = readback.external_delivery
    if not recovery or identity is None or not isinstance(plan, dict):
        raise RuntimeError("TypeScript refresh admission result missing")
    decision = recovery["decision"]
    delivery_error = plan.get("error_code") if decision != "reject" else None
    if decision not in {"replay", "repair_receipt", "reject"} and not delivery_error:
        # Record a requested pause before a new writeback can commit. If later
        # validation fails, retaining the pause is conservative and retryable.
        if (plan.get("transition") or {}).get("state") == "paused":
            finish_external_delivery_refresh({"ok": True}, readback, runtime_root, dry_run=dry_run)
        return None
    payload = {
        **(readback.writeback_run or {}), "ok": decision != "reject" and not delivery_error,
        "dry_run": dry_run, "appended": False,
        "idempotent_replay": decision == "replay" and not delivery_error,
        "receipt_repair_required": decision == "repair_receipt" and not dry_run,
        "registry": str(registry_path), "runtime_root": str(runtime_root), "goal_id": goal_id,
        "refresh_recovery": recovery, "settlement_identity": identity.as_dict(),
        "settlement_result": settlement_result_payload(readback.delivery),
    }
    if decision == "reject":
        payload["error"] = (
            f"{recovery['reason']}: committed writeback is unchanged; "
            "do not begin a new Turn or repeat spend to repair it. "
            "Retry the original delivery fields with only the missing vision decision; "
            "if a newer vision already exists, inspect current quota instead."
        )
    elif delivery_error:
        payload["error_code"] = delivery_error
        payload["refresh_recovery"] = {**recovery, "decision": "reject", "reason": delivery_error}
        key = plan.get("resume_key")
        payload["error"] = (
            f"{delivery_error}: external delivery was not resumed. "
            "Keep --suppress-external-sinks for local recovery. "
            + (f"To deliberately resume this operation, retry the same recovery command with "
               f"--resume-external-sinks {key} instead of --suppress-external-sinks. " if key else "")
            + "Existing provider permissions still apply; do not repeat business mutations or spend."
        )
    return finish_external_delivery_refresh(payload, readback, runtime_root, dry_run=dry_run)
